### ISSUE TITLE
`@types/npmcli__package-json` should be a `dependency`, not `devDependency`

### ISSUE THREAD

**Issue by erheron** [webpro/knip#283]
Hi 👋🏻 
I installed `knip` in our project, and discovered that we need to manually add `@types/npmcli__package-json` to our dependencies in `package.json`.  Otherwise, Typescript fails. 

That's because `knip` specifies it as a `devDependency`, and `devDependencies` are not automatically installed.
Also, it is  a real dependency, since `knip` exports types which use types from `@types/npmcli__package-json` under the hood.

Shall it be converted to a `dependency`?

**Comment by webpro** [comment]
Not sure I follow. Do you mean that Knip is exporting something from `@npmcli/package-json`?

Are you programmatically using and importing `knip`? That's not supported/documented (yet).

**Comment by erheron** [comment]
> Not sure I follow. Do you mean that Knip is exporting something from @npmcli/package-json?

Yep, `knip` is transitively using types from the repo I mentioned. 

> Are you programmatically using and importing knip? That's not supported/documented (yet).

Not sure what exactly you mean by "programmatically" 🤷🏻  I was trying to use `knip` according to the docs, where it is mentioned that one can use `knip.ts` as a configuration file. 

Here is a reproduction: https://github.com/erheron/knip-issue-283-reproduction/

**Comment by webpro** [comment]
You mention `yarn tsc`, but `tsc` is the TypeScript compiler. You don't need to compile the `knip.ts` file.

The `knip.ts` configuration file is read and used by Knip as a dynamic alternative to `knip.json`. The former can be typed by importing `KnipConfig` from Knip, the latter can be validated using `"$schema": "https://unpkg.com/knip@2/schema.json"`. Using a dynamic `knip.ts` file could be a preference over JSON, but it's required in certain situations, e.g. when using compilers.

The topics you mention are documented starting around https://github.com/webpro/knip#lets-go

You can run `yarn knip` to lint/analyze your code, while `yarn tsc` can be used to compile/build your code.

**Comment by erheron** [comment]
Hmm looks like my previous explanation was a bit unclear 😂 Let me try again 👇🏻 

1. In the docs, it is mentioned that one can use `knip.ts` as a configuration file
2. To me, one of the main reasons to do that is that we can employ Typescript to perform __compile-time validation of that configuration file__ for us 🤖 
3. To have that compile-time validation, we have to compile that file 🤷🏻 
4. As the reproduction shows, that is not possible (with default setup). One can, of course, manually add `@types/npmcli__package-json` to projec't dependencies, but what I'm trying to say here is that this should happen "automatically" i.e. it should be moved from knip's `devDependencies` to `dependencies`

**Comment by webpro** [comment]
You don't need compile-time validation, since you don't need to compile it. It's not part of your program. What you want cq what Knip provides is type-safety when authoring the configuration file. The value is mostly during author-time type hints. The `KnipConfig` provides all the types you need. The types eventually coming from `@npmcli/package-json` aren't relevant to your `knip.ts`.

The actual process of compiling your `knip.ts` alone does not define the ability to use TypeScript in said file. The language server in VS Code and other editors analyze the code, but they don't (need to) compile and emit it.

However, you still have a point in saying that this should be possible. See #248 for more details and discussion to that specifically. Personally I prefer to keep `devDependencies` out of the published package, perhaps I can look into duck typing as @danvk suggests over there.

**Comment by erheron** [comment]
>  What you want cq what Knip provides is type-safety when authoring the configuration file. The value is mostly during author-time type hints.

Right, right. I agree that language servers in different editors are helping here. But ultimately there's no better tool than just add a file to compile target and being 100% sure it's "correct". 


> However, you still have a point in saying that this should be possible. See https://github.com/webpro/knip/issues/248 for more details and discussion to that specifically.

Yeah I saw that issue before, but felt that it's a bit different (#248 is about how knip handles this situation in other packages, this issue is about knip itself 😂 )

I'm OK if we close this and leave a reference to #248  🙌🏻

**Comment by webpro** [comment]
That's indeed a different issue, but essentially shares the same topic: (dev) dependency bloat.

Let's keep this one open until there's a fix. Feel free to pick it up yourself. I'll update the board as needed when I pick it up myself.

**Comment by webpro** [comment]
Upon further thought, going to close this one. The issue is mainly that it's not only `@types/npmcli__package-json`, and there will be more in the future. It's a broader issue in the ecosystem and I'm not sure what's the best way to go about it atm.

I'll reconsider when there's better solutions out there and/or when importing the main routine is supported.
